src/Strategy/Maven/Pom.hs

```diff
@@ -142,16 +143,39 @@ reifyDeps pom = M.mapWithKey overlayDepManagement (pomDependencies pom)
     overlayDepManagement :: (Group, Artifact) -> MvnDepBody -> MvnDepBody
     overlayDepManagement key body = maybe body (body <>) (M.lookup key (pomDependencyManagement pom))
 
--- Naively interpolate properties into a Text. This only interpolates Text that
--- starts with "${" and ends with "}", e.g.,
---     "${myproperty}"
--- will interpolate the value of "myproperty", but
---     "blah-${myproperty}"
--- will not have its property interpolated
-naiveInterpolate :: Map Text Text -> Text -> Text
-naiveInterpolate properties text
-  | T.isPrefixOf "${" text,
-    T.isSuffixOf "}" text =
-    let stripped = T.drop 2 (T.init text)
-     in fromMaybe ("PROPERTY NOT FOUND: " <> text) (M.lookup stripped properties)
+-- | Interpolate Pom properties into a string with the ${property} format This
+-- interpolates both computed/built-in properties and user-specified properties,
+-- preferring user-specified properties
+interpolateProperties :: Pom -> Text -> Text
+interpolateProperties pom = interpolate (pomProperties pom <> computeBuiltinProperties pom)
+
+-- | Compute the most-commonly-used builtin properties for package resolution
+computeBuiltinProperties :: Pom -> Map Text Text
+computeBuiltinProperties pom = M.fromList
+  [ ("project.groupId", coordGroup (pomCoord pom))
+  , ("project.artifactId", coordArtifact (pomCoord pom))
+  , ("project.version", coordVersion (pomCoord pom))
+  ]
+
+interpolate :: Map Text Text -> Text -> Text
+interpolate properties text
+  | Just (beforeBegin, afterBegin) <- breakOnAndRemove "${" text
+  , Just (property, afterEnd) <- breakOnAndRemove "}" afterBegin =
```

Comment:
Not sure I understand this syntax here.  Can you explain this?

Reply:
so normally guards are used for boolean values:

```
foo thing
  | someCondition thing
  , andSomeOtherCondition thing = ...
  | otherwise = ...
```

but there's also something called a "pattern guard", which is essentially just pattern matching as a guard. in this case, I'm pattern matching that the results of both of these `breakOnAndRemove` are `Just ...`. If that condition fails for either one, it'll fall through to the next guard (the `| otherwise =`)

Reply:
I think this might be more readable as a function on its own:
```haskell
splitPropertyString :: Text -> Maybe (Text, Text, Text)
splitPropertyString tx = _
```

where `splitPropertyString "a${b}c"` would return `Just ("a", "b", "c")`.